[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors

anshul-si opened the pull request:
**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170632
* __->__ #170040
* #170035
* #170030



cc @wanchaol @tianyu-l @wz337 @XilunWu @d4l3k @pragupta @SherlockNoMad @ppwwyyxx

anshul-si committed:
[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"




[ghstack-poisoned]

wconstab reviewed (approved):
lgtm. you can probably rebase/reorder this one together with the first one and land them while we work on the other 2

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si commented:
@pytorchbot revert -m "regression in sixlib" -c nosignal

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] fixes unnecessary redistributions when subtracting two partial tensors"

**Summary:** Currently, whenever we subtract two partial dtensors, we redistribute since linearity is -1 for aten.sub.tensor. However, this is an unnecessary redistribution that can be avoided in similar ways to its add counterpart. I moved the op to linear_ops and ensured subtracting a scalar from a partial dtensor continues to redistribute.

**Test Cases:**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_sub_scalar_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]
